libflux/flux-core/src/scanner/scanner.rl

```diff
@@ -46,7 +46,9 @@ use crate::scanner::*;
     escaped_char = "\\" ( "n" | "r" | "t" | "\\" | '"' | "${" );
     unicode_value = (any_count_line - [\\$]) | escaped_char;
     byte_value = "\\x" xdigit{2};
-    dollar_value = "$" ( any_count_line - "{" );
+    # when the dollar is in the last position before the closing quotes
+    # we need to match the quote char but not consume it.
+    dollar_value = "$" ( any_count_line - [{"] | "\"" %{ fhold; } );
     string_lit_char = ( unicode_value | byte_value | dollar_value );
     string_lit = '"' string_lit_char* "$"? :> '"';
```

Comment:
The dollar is already allowed at the end of `string_lit`, so permitting it in `string_lit_char` is redundant. The other thing about this approach is that using an `fhold` in a grammar fragment that is included in other places is risky, because the equivalent state machine is no longer consuming the strings specified by the grammer. It then becomes hard to reason about. The `fhold` trick should be used sparingly, and generally at the top-most level of a grammar.

It's correct to exclude `$"` from dollar value though, so I would leave that there. That would look like:

```
dollar_value = "$" ( any_count_line - [{"] );
```

To get correct handling of `"ab$"` from the parser, instead I would permit dollar at the end of a string in the `string_expr` scanner by adding the following:

```
'$"' => { fhold; te -= 1; tok = TokenType::Text; fbreak; };
```

By doing it this way, the scanner pattern for string expressions and the more detailed string expression scanner are taking the same approach: allow a single dollar at the end. We have to add `te -= 1` here because the text matched is taken from `te` (token end), which has already been derived from `p` at this point. So we need to update it too.

Reply:
makes sense; thanks!

Reply:
I'll take another stab; something is not quite working out